src/ucp/tag/tag_send.c

```diff
@@ -264,9 +264,12 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nbx,
                                              contig_length, tag);
             ucp_request_send_check_status(status, ret, goto out);
         }
-    } else {
+    } else if (attr_mask == UCP_OP_ATTR_FLAG_NO_IMM_CMPL) {
         datatype      = ucp_dt_make_contig(1);
         contig_length = count;
+    } else {
+        datatype      = param->datatype;
+        contig_length = UCP_DT_IS_CONTIG(datatype) ? count : 0;
```

Comment:
need to use `ucp_contig_dt_length(datatype, count);` here

Reply:
`contig_length = UCP_DT_IS_CONTIG(datatype) ? ucp_contig_dt_length(datatype, count) : 0;` ?

Reply:
but can it happen that `datatype` is not specified here?
`UCP_OP_ATTR_FIELD_DATATYPE` wasn't set

Reply:
no this branch means that both flags are set

Reply:
> no, this branch means that both flags are set

maybe it worth adding `ucs_assert()`?

Reply:
what kind?

Reply:
> what kind?

@brminich `ucs_assert(attr_mask & (UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FLAG_NO_IMM_CMPL));`

Reply:
> need to use ucp_contig_dt_length(datatype, count); here
yep, copy-paste issue

added assert

Reply:
imo checking branch condonation in asserts is too excessive

Reply:
removed assert